src/ffi/datetime.rs

```diff
@@ -288,6 +347,7 @@ macro_rules! _access_field {
 
 // Accessor functions for PyDateTime_Date and PyDateTime_DateTime
 #[inline]
+#[cfg(not(PyPy))]
```

Comment:
For the accessor functions, might it make sense to do this in two blocks like this:

```rust
#[cfg(not(PyPy))]
{
...  // Current code
}

#[cfg(PyPy)] {
 #[link_name = "PyPyDateTime_DATE_GET_HOUR"]
    pub fn PyDateTime_DATE_GET_HOUR(o: *mut PyObject) -> c_int;
    #[link_name = "PyPyDateTime_DATE_GET_MICROSECOND"]
    pub fn PyDateTime_DATE_GET_MICROSECOND(o: *mut PyObject) -> c_int;
    #[link_name = "PyPyDateTime_DATE_GET_MINUTE"]
    ...
}

Seems like something that would be better than individually annotating every function.

Reply:
If we wanted to do this, we would need something like `cfg-if` (since we cant just arbitrarly `#[cfg]` blocks on this scope).

![image](https://user-images.githubusercontent.com/2467993/54948882-66a67200-4f46-11e9-9279-567f25aabd07.png)

Do you think it's worth adding for this case?

Reply:
For something like this I think either `cfg-if` or a custom macro would be worth it, but I'll defer to whatever @konstin thinks.

Another possibility is that I think we don't actually *need* to call the PyPy's implementation of these macros, since in this case we're not actually making calls to the C API, we're just *replicating* the C API in Rust. I'm pretty sure we can just leave all of this stuff alone.

Reply:
This is why I've ignored all the other macros (`PyDateTime_CheckExact` for example) in many other places, because they're shallow enough so they behave the same.

But `PyDateTime_GET_YEARS` for example does seem to access some internals of structs, so I wasn't sure that it's completely OK to assume it will behave the same.

Reply:
@omerbenamram Hm.. That's a good point. I am not 100% clear about this, but I *think* you already are making these assumptions because the structures [are defined in Rust in `ffi/datetime.rs`](https://github.com/PyO3/pyo3/blob/27fb0f3fd0c685d7a72a52239d605e9fbeacf5e2/src/ffi/datetime.rs#L155), and the "macros" are operating on these internal structures.

It seems like there would be many other problems if the C representations of these objects is different depending on whether you constructed it in Rust or in PyPy.

I think it's probably safe to remove the PyPy-specific stuff here, but I'm willing to be convinced otherwise.

Reply:
It does break the tests (good thing that there are tests :)!

This is with removing the PyPy (`PyDateTime_GET_*`) macros.

```
=================================================================================================================================== FAILURES ====================================================================================================================================
______________________________________________________________________________________________________________________________ test_date_accessors ______________________________________________________________________________________________________________________________

    @given(d=st.dates())
>   def test_date_accessors(d):

tests/test_datetime.py:73:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

d = datetime.date(2000, 1, 1)

    @given(d=st.dates())
    def test_date_accessors(d):
        act = rdt.get_date_tuple(d)
        exp = (d.year, d.month, d.day)

>       assert act == exp
E       assert (0, 0, 0) == (2000, 1, 1)
E         At index 0 diff: 0 != 2000
E         Use -v to get the full diff

tests/test_datetime.py:77: AssertionError
---------------------------------------------------------------------------------------------------------------------------------- Hypothesis -----------------------------------------------------------------------------------------------------------------------------------
Falsifying example: test_date_accessors(d=datetime.date(2000, 1, 1))
___________________________________________________________________________________________________________________________________ test_time ___________________________________________________________________________________________________________________________________

    @given(t=st.times())
>   def test_time(t):

tests/test_datetime.py:112:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

t = datetime.time(0, 0, 0, 1)

    @given(t=st.times())
    def test_time(t):
        act = rdt.get_time_tuple(t)
        exp = (t.hour, t.minute, t.second, t.microsecond)

>       assert act == exp
E       assert (0, 0, 0, 0) == (0, 0, 0, 1)
E         At index 3 diff: 0 != 1
E         Use -v to get the full diff

tests/test_datetime.py:116: AssertionError
---------------------------------------------------------------------------------------------------------------------------------- Hypothesis -----------------------------------------------------------------------------------------------------------------------------------
Falsifying example: test_time(t=datetime.time(0, 0, 0, 1))
______________________________________________________________________________________________________________________________ test_datetime_tuple ______________________________________________________________________________________________________________________________

    @given(dt=st.datetimes())
>   def test_datetime_tuple(dt):

tests/test_datetime.py:194:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

dt = datetime.datetime(2000, 1, 1, 0, 0)

    @given(dt=st.datetimes())
    def test_datetime_tuple(dt):
        act = rdt.get_datetime_tuple(dt)
        exp = dt.timetuple()[0:6] + (dt.microsecond,)

>       assert act == exp
E       assert (0, 0, 0, 0, 0, 0, ...) == (2000, 1, 1, 0, 0, 0, ...)
E         At index 0 diff: 0 != 2000
E         Use -v to get the full diff

tests/test_datetime.py:198: AssertionError
---------------------------------------------------------------------------------------------------------------------------------- Hypothesis -----------------------------------------------------------------------------------------------------------------------------------
Falsifying example: test_datetime_tuple(dt=datetime.datetime(2000, 1, 1, 0, 0))
_____________________________________________________________________________________________________________________________ test_delta_accessors ______________________________________________________________________________________________________________________________

    @given(td=st.timedeltas())
>   def test_delta_accessors(td):

tests/test_datetime.py:264:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

td = datetime.timedelta(0, 0, 1)

    @given(td=st.timedeltas())
    def test_delta_accessors(td):
        act = rdt.get_delta_tuple(td)
        exp = (td.days, td.seconds, td.microseconds)

>       assert act == exp
E       assert (1, 0, 0) == (0, 0, 1)
E         At index 0 diff: 1 != 0
E         Use -v to get the full diff

tests/test_datetime.py:268: AssertionError
---------------------------------------------------------------------------------------------------------------------------------- Hypothesis -----------------------------------------------------------------------------------------------------------------------------------
Falsifying example: test_delta_accessors(td=datetime.timedelta(0, 0, 1))
===================================================================================================== 4 failed, 38 passed, 3 skipped, 6 xfailed, 9 xpassed in 2.35 seconds ======================================================================================================
```

Reply:
Hm.. Interesting. Does this mean that the various datetime structs we define are unused in PyPy?

In any case, if it's true that we need these alternate macro definitions, I think maybe something like `[cfg-if]` will be nicer and easier to maintain.

@konstin Thoughts?

Reply:
I don't think cfg-if works in that scope, and I don't have a problem with annotating functions individually. However it would be possible to group those functions into modules (that are re-exported from other modules) and then annotate the modules, though I don't know if that's worth it.